### PR TITLE
Fix limited set issues

### PR DESCRIPTION
**Pull request by kevinhyang**
@rickhutcheson @udemy/platform 

https://udemyjira.atlassian.net/browse/EP-990

### CODE REVIEW

**Review by jpaulodit — APPROVED** [review]

**Review by rickhutcheson — DISMISSED** [review]

**Review by rickhutcheson — CHANGES_REQUESTED** [review]
I don't think we can raise here. This will probably cause workers to die

**Comment by kevinhyang** [comment]
@rickhutcheson PTAL, I've revised the error into a Signal. I'm curious as to what used to be in the spot of "I was wrong" 😂

**Comment by kevinhyang** [comment]
Modified the above slightly because the Signal base's `def send()` expects sender to either be an object or None.

File: celery/datastructures.py [hunk 1]
```diff
@@ -552,120 +553,220 @@ def values(self):
...
+        """
+        if not now:
+            now = time.time()
+        if hasattr(now, '__call__'):
+            now = now()  # if we got this now as function, evaluate it
+        if self.maxlen:
+            while len(self._data) > self.maxlen:
+                inserted_time, _ = self._heap[0]
+                if inserted_time + self.expires > now:
+                    signals.limited_set_purging_unexpired_members.send(sender=None)
```
Comment (rickhutcheson) [inline celery/datastructures.py]: shouldn't we only send it once? or once / task type?

**Comment by jpaulodit** [comment]
Are we making code changes that produces different behavior from original code? Is the plan to maintain our own fork long term?

**Comment by kevinhyang** [comment]
@jpaulodit For now, yes. We've had our own fork of Celery for a few months now.

**Comment by jpaulodit** [comment]
Can you add something about the branch that we're using for our changes. And maybe short reason why, or a link to documentation that explains why. There needs to be something on the main page that tells a new engineer something. Thanks.

**Comment by kevinhyang** [comment]
There's information in CHANGELOG.rst that I've been updating. Should I a link to that somewhere?

File: celery/datastructures.py [hunk 2]
```diff
@@ -695,10 +695,10 @@ def purge(self, now=None):
         if hasattr(now, '__call__'):
             now = now()  # if we got this now as function, evaluate it
         if self.maxlen:
+            inserted_time, _ = self._heap[0]
```
Comment (rickhutcheson) [inline celery/datastructures.py]: this will have an error if self._data is empty

Comment (kevinhyang) [reply celery/datastructures.py]: I'm missing an if condition....

**Review by rickhutcheson — APPROVED** [review]

**Pull request merged**